### PR TITLE
Add MASTG-TEST-0231 for weak encryption modes on Android

### PR DESCRIPTION
**Pull request by cpholguera**

### CODE REVIEW

File: tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0231.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+weakness: MASWE-0020
+---
+
+## Overview
+
+To test for the [use of weak encryption modes](../../../Document/0x04g-Testing-Cryptography.md#weak-block-cipher-mode) in Android apps, we need to focus on methods from cryptographic frameworks and libraries that are used to configure and apply encryption modes.
+
+In Android development, the `Cipher` class from the Java Cryptography Architecture (JCA) is the primary API that allows you to specify the encryption mode for cryptographic operations. [`Cipher.getInstance`](https://developer.android.com/reference/javax/crypto/Cipher#getInstance(java.lang.String)) defines the transformation string, which includes the encryption algorithm, mode of operation, and padding scheme. The general format is `"Algorithm/Mode/Padding"`. For example:
+
+```java
```
Comment (serek8) [inline tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0231.md:15]: no kotlin?

File: tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0231.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
+```
```
Comment (cpholguera) [inline tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0231.md:17]:
````suggestion
```kotlin
Cipher.getInstance("AES/CBC/PKCS5Padding")
```
````

**Comment by serek8** [comment]
@cpholguera Can we also include a set of weak `Algorithm/Mode/Padding` that a tester can check for? Right now there are only two examples, but I think there are more, right?

**Comment by cpholguera** [comment]
We could but we haven't been including those so explicitly until now so I'd leave it in the current form and we can revisit this and others later once we decide how to enter that kind of references.

File: tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0231.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+In this test we're going to focus on symmetric encryption modes such as [ECB (Electronic Codebook)](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Electronic_codebook_(ECB)) operate deterministically, dividing plaintext into blocks and encrypting them separately, which reveals patterns in the ciphertext. This makes it vulnerable to attacks like [known-plaintext attacks](https://en.wikipedia.org/wiki/Known-plaintext_attack) and [chosen-plaintext attacks](https://en.wikipedia.org/wiki/Chosen-plaintext_attack).
+
+**Out of Scope**: Asymmetric encryption modes like RSA are out of scope for this test because they don't use block modes like ECB.
+
+In the transformation strings like `"RSA/ECB/OAEPPadding"` or "RSA/ECB/PKCS1Padding", the inclusion of "ECB" in this context is misleading. Unlike symmetric ciphers, **RSA doesn't operate in block modes like ECB**. The "ECB" designation is a [placeholder in some cryptographic APIs](https://github.com/openjdk/jdk/blob/680ac2cebecf93e5924a441a5de6918cd7adf118/src/java.base/share/classes/com/sun/crypto/provider/RSACipher.java#L126) and doesn't imply that RSA uses ECB mode. Understanding these nuances helps prevent false positives.
```
Comment (cpholguera) [inline tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0231.md:23]:
```suggestion
In the transformation strings like "RSA/ECB/OAEPPadding" or "RSA/ECB/PKCS1Padding", the inclusion of "ECB" in this context is misleading. Unlike symmetric ciphers, **RSA doesn't operate in block modes like ECB**. The "ECB" designation is a [placeholder in some cryptographic APIs](https://github.com/openjdk/jdk/blob/680ac2cebecf93e5924a441a5de6918cd7adf118/src/java.base/share/classes/com/sun/crypto/provider/RSACipher.java#L126) and doesn't imply that RSA uses ECB mode. Understanding these nuances helps prevent false positives.
```

File: tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0231.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+## Observation
+
+The output should contain a list of locations where insecure or deprecated encryption modes are used in cryptographic operations.
+
+## Evaluation
+
+The test case fails if any insecure encryption modes are identified in the app.
+
+Replace insecure encryption modes with secure block cipher modes such as [AES-GCM or AES-CCM](https://csrc.nist.gov/pubs/sp/800/38/d/final) which are authenticated encryption modes that provide confidentiality, integrity, and authenticity.
```
Comment (cpholguera) [inline tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0231.md:37]:
```suggestion

## Mitigation

Replace insecure encryption modes with secure block cipher modes such as [AES-GCM or AES-CCM](https://csrc.nist.gov/pubs/sp/800/38/d/final) which are authenticated encryption modes that provide confidentiality, integrity, and authenticity.
```

File: tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0231.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+## Observation
+
+The output should contain a list of locations where insecure or deprecated encryption modes are used in cryptographic operations.
+
+## Evaluation
+
+The test case fails if any insecure encryption modes are identified in the app.
+
+
```
Comment (cpholguera) [inline tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0231.md:37]:
```suggestion
```

**Review by serek8 — APPROVED** [review]

File: tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0231.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+In Android development, the `Cipher` class from the Java Cryptography Architecture (JCA) is the primary API that allows you to specify the encryption mode for cryptographic operations. [`Cipher.getInstance`](https://developer.android.com/reference/javax/crypto/Cipher#getInstance(java.lang.String)) defines the transformation string, which includes the encryption algorithm, mode of operation, and padding scheme. The general format is `"Algorithm/Mode/Padding"`. For example:
+
+```kotlin
+Cipher.getInstance("AES/CBC/PKCS5Padding")
+```
+
+In this test we're going to focus on symmetric encryption modes such as [ECB (Electronic Codebook)](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Electronic_codebook_(ECB)) operate deterministically, dividing plaintext into blocks and encrypting them separately, which reveals patterns in the ciphertext. This makes it vulnerable to attacks like [known-plaintext attacks](https://en.wikipedia.org/wiki/Known-plaintext_attack) and [chosen-plaintext attacks](https://en.wikipedia.org/wiki/Chosen-plaintext_attack).
```
Comment (cpholguera) [inline tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0231.md:19]:
````suggestion
In this test we're going to focus on symmetric encryption modes such as [ECB (Electronic Codebook)](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Electronic_codebook_(ECB)) operate deterministically, dividing plaintext into blocks and encrypting them separately, which reveals patterns in the ciphertext. This makes it vulnerable to attacks like [known-plaintext attacks](https://en.wikipedia.org/wiki/Known-plaintext_attack) and [chosen-plaintext attacks](https://en.wikipedia.org/wiki/Chosen-plaintext_attack).

For example, the following transformations are all [considered vulnerable](https://support.google.com/faqs/answer/10046138?hl=en):

- "AES" (uses AES/ECB mode by default)
- "AES/ECB/NoPadding"
- "AES/ECB/PKCS5Padding"
- "AES/ECB/ISO10126Padding"

You can learn more about ECB and other modes in [NIST SP 800-38A - Recommendation for Block Cipher Modes of Operation: Methods and Techniques](https://csrc.nist.gov/pubs/sp/800/38/a/final). Also check the [Decision to Revise NIST SP 800-38A, Recommendation for Block Cipher Modes of Operation: Methods and Techniques](https://csrc.nist.gov/news/2023/decision-to-revise-nist-sp-800-38a) and [NIST IR 8459
Report on the Block Cipher Modes of Operation in the NIST SP 800-38 Series](https://nvlpubs.nist.gov/nistpubs/ir/2024/NIST.IR.8459.pdf) for the latest information. 
````

File: tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0231.md [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+---
+
+## Overview
+
+To test for the [use of weak encryption modes](../../../Document/0x04g-Testing-Cryptography.md#weak-block-cipher-mode) in Android apps, we need to focus on methods from cryptographic frameworks and libraries that are used to configure and apply encryption modes.
+
+In Android development, the `Cipher` class from the Java Cryptography Architecture (JCA) is the primary API that allows you to specify the encryption mode for cryptographic operations. [`Cipher.getInstance`](https://developer.android.com/reference/javax/crypto/Cipher#getInstance(java.lang.String)) defines the transformation string, which includes the encryption algorithm, mode of operation, and padding scheme. The general format is `"Algorithm/Mode/Padding"`. For example:
+
+```kotlin
+Cipher.getInstance("AES/CBC/PKCS5Padding")
```
Comment (cpholguera) [inline tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0231.md:16]:
```suggestion
Cipher.getInstance("AES/ECB/PKCS5Padding")
```

File: tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0231.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+**Out of Scope**: Asymmetric encryption modes like RSA are out of scope for this test because they don't use block modes like ECB.
+
+In the transformation strings like `"RSA/ECB/OAEPPadding"` or `"RSA/ECB/PKCS1Padding"`, the inclusion of "ECB" in this context is misleading. Unlike symmetric ciphers, **RSA doesn't operate in block modes like ECB**. The `ECB` designation is a [placeholder in some cryptographic APIs](https://github.com/openjdk/jdk/blob/680ac2cebecf93e5924a441a5de6918cd7adf118/src/java.base/share/classes/com/sun/crypto/provider/RSACipher.java#L126) and doesn't imply that RSA uses ECB mode. Understanding these nuances helps prevent false positives.
```
Comment (cpholguera) [inline tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0231.md:23]:
```suggestion
In the transformation strings like `"RSA/ECB/OAEPPadding"` or `"RSA/ECB/PKCS1Padding"`, the inclusion of `ECB` in this context is misleading. Unlike symmetric ciphers, **RSA doesn't operate in block modes like ECB**. The `ECB` designation is a [placeholder in some cryptographic APIs](https://github.com/openjdk/jdk/blob/680ac2cebecf93e5924a441a5de6918cd7adf118/src/java.base/share/classes/com/sun/crypto/provider/RSACipher.java#L126) and doesn't imply that RSA uses ECB mode. Understanding these nuances helps prevent false positives.
```

File: tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0231.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+For example, the following transformations are all [considered vulnerable](https://support.google.com/faqs/answer/10046138?hl=en):
+
+- "AES" (uses AES/ECB mode by default)
+- "AES/ECB/NoPadding"
+- "AES/ECB/PKCS5Padding"
+- "AES/ECB/ISO10126Padding"
+
+You can learn more about ECB and other modes in [NIST SP 800-38A - Recommendation for Block Cipher Modes of Operation: Methods and Techniques](https://csrc.nist.gov/pubs/sp/800/38/a/final). Also check the [Decision to Revise NIST SP 800-38A, Recommendation for Block Cipher Modes of Operation: Methods and Techniques](https://csrc.nist.gov/news/2023/decision-to-revise-nist-sp-800-38a) and [NIST IR 8459
+Report on the Block Cipher Modes of Operation in the NIST SP 800-38 Series](https://nvlpubs.nist.gov/nistpubs/ir/2024/NIST.IR.8459.pdf) for the latest information.
```
Comment (cpholguera) [inline tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0231.md:29]:
```suggestion
You can learn more about ECB and other modes in [NIST SP 800-38A - Recommendation for Block Cipher Modes of Operation: Methods and Techniques](https://csrc.nist.gov/pubs/sp/800/38/a/final). Also check the [Decision to Revise NIST SP 800-38A, Recommendation for Block Cipher Modes of Operation: Methods and Techniques](https://csrc.nist.gov/news/2023/decision-to-revise-nist-sp-800-38a) and [NIST IR 8459 Report on the Block Cipher Modes of Operation in the NIST SP 800-38 Series](https://nvlpubs.nist.gov/nistpubs/ir/2024/NIST.IR.8459.pdf) for the latest information.
```

File: tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0231.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+```kotlin
+Cipher.getInstance("AES/ECB/PKCS5Padding")
+```
+
+In this test we're going to focus on symmetric encryption modes such as [ECB (Electronic Codebook)](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Electronic_codebook_(ECB)) operate deterministically, dividing plaintext into blocks and encrypting them separately, which reveals patterns in the ciphertext. This makes it vulnerable to attacks like [known-plaintext attacks](https://en.wikipedia.org/wiki/Known-plaintext_attack) and [chosen-plaintext attacks](https://en.wikipedia.org/wiki/Chosen-plaintext_attack).
+
+For example, the following transformations are all [considered vulnerable](https://support.google.com/faqs/answer/10046138?hl=en):
+
+- "AES" (uses AES/ECB mode by default)
```
Comment (cpholguera) [inline tests-beta/android/MASVS-CRYPTO/MASTG-TEST-0231.md:23]:
```suggestion
- "AES" (uses AES/ECB mode by [default](https://docs.oracle.com/javase/6/docs/technotes/guides/security/crypto/CryptoSpec.html#Cipher))
```

**Pull request merged**